src/servers/src/mysql/handler.rs

```diff
@@ -211,3 +261,96 @@ impl<W: AsyncWrite + Send + Sync + Unpin> AsyncMysqlShim<W> for MysqlInstanceShi
         w.ok().await.map_err(|e| e.into())
     }
 }
+
+fn replace_params(params: Vec<ParamValue>, query: String) -> String {
+    let mut query = query;
+    let mut index = 1;
+    for param in params {
+        let s = match param.value.into_inner() {
+            ValueInner::Int(u) => u.to_string(),
+            ValueInner::UInt(u) => u.to_string(),
+            ValueInner::Double(u) => u.to_string(),
+            ValueInner::NULL => "NULL".to_string(),
+            ValueInner::Bytes(b) => format!("'{}'", &String::from_utf8_lossy(b)),
+            ValueInner::Date(_) => NaiveDate::from(param.value).to_string(),
+            ValueInner::Datetime(_) => NaiveDateTime::from(param.value).to_string(),
+            ValueInner::Time(_) => format_duration(Duration::from(param.value)),
+        };
+        query = query.replace(&format!("${}", index), &s);
+        index += 1;
+    }
+    query
+}
+
+fn format_duration(duration: Duration) -> String {
+    let seconds = duration.as_secs() % 60;
+    let minutes = (duration.as_secs() / 60) % 60;
+    let hours = (duration.as_secs() / 60) / 60;
+    format!("{}:{}:{}", hours, minutes, seconds)
+}
+
+async fn validate_query(query: &str) -> Result<Statement> {
+    let statement = ParserContext::create_with_dialect(query, &GenericDialect {});
+    let mut statement = match statement {
+        Err(e) => {
+            return Err(Error::PrepareStatementFailed {
+                err_msg: e.to_string(),
+            })
+        }
+        Ok(s) => s,
+    };
+
+    if statement.len() != 1 {
+        return Err(Error::PrepareStatementFailed {
+            err_msg: "prepare statement only support single statement".to_string(),
+        });
+    }
+
+    let statement = statement.remove(0);
+
+    match statement {
+        Statement::Query(_) => Ok(statement),
+        _ => Err(Error::PrepareStatementFailed {
+            err_msg: "prepare statement only support SELECT now".to_string(),
+        }),
+    }
+}
+
+async fn write_output<'a, W: AsyncWrite + Send + Sync + Unpin>(
+    w: QueryResultWriter<'a, W>,
+    query: &str,
+    outputs: Vec<Result<Output>>,
+) -> Result<()> {
+    let mut writer = MysqlResultWriter::new(w);
+    for output in outputs {
+        writer.write(query, output).await?;
+    }
+    Ok(())
+}
+
+// dummy columns to satisfy opensrv_mysql
+// just the number of params is useful
+fn dummy_params(index: u32) -> Vec<Column> {
```

Comment:
Here we can add a TODO to use parameter type inference. Maybe some mysql clients will use this data just like the situation in postgresql, rust-postgres relies on server type inference for parameter and column types

Reply:
@sunng87  Got it. My first implementation is the dafafusion approach https://github.com/GreptimeTeam/greptimedb/compare/develop...SSebo:greptimedb:prepare_datafusion, I'll continue the work. 

Should I return the `LogicalPlan` in `do_describe` as second parameter?
`fn describe(&self, stmt: QueryStatement, query_ctx: QueryContextRef) -> Result<(Schema, LogicalPlan)>;`

Reply:
I'm not sure if it's possible at the moment. It requires refactoring to bring query engine into protocol server. @MichaelScofield is planning an overhaul on design of protocol servers. I think eventually we should cache LogicalPlan as prepared statement.